autoprotocol/instruction.py

```diff
@@ -86,6 +100,119 @@ def filter_criteria(item):
             ]
         return data
 
+    def _set_informatics(self):
+        """
+        set Instruction informatics to valid list of Informatics
+
+        Raises
+        ------
+        TypeError
+            informatics must be a list
+        TypeError
+            informatics element must be a dict
+        KeyError
+            informatics must have keys 'type' and 'data'
+        ValueError
+            informatics type must be a valid type
+        """
+        # any informatics dict must only have these keys.
+        valid_informatics_keys = ["type", "data"]
+        # informatics must be a list of dict.
+        informatics = self.informatics
+        if isinstance(informatics, list):
+            for info in informatics:
+                if not isinstance(info, dict):
+                    raise TypeError(f"informatics: {info} must be provided in a dict.")
+        else:
+            raise TypeError(
+                f"informatics: {informatics} must be provided in a list of dict. "
+            )
+
+        # instantiate Informatics based on the type
+        valid_informatics = []
+        for info in self.informatics:
+            info_keys = info.keys()
+            if set(info_keys) != set(valid_informatics_keys):
+                raise KeyError(
+                    f"informatics dict: {info} must have keys: {valid_informatics_keys}"
+                )
+            # validate wells only if `wells` param is present in informatics
+            try:
+                if info["data"]["wells"]:
+                    self._check_info_wells()
+            except KeyError:
+                continue
+            info_type = info["type"]
+            if info_type == "attach_compounds":
+                info_obj = AttachCompounds(info["data"])
+            else:
+                raise ValueError(f"informatics type: {info_type} is not a valid type.")
+            valid_informatics.append(info_obj)
+
+        self.informatics = valid_informatics
+
+    def _check_info_wells(self):
+        """
+        validates Informatics wells are included in the wells associated with
+        the instruction.
+
+        Raises
+        -------
+        TypeError
+            wells are Well, list of Well or WellGroup
+        ValueError
+            Informatics wells are part of wells Instruction is operating on
+        """
+        for info in self.informatics:
```

Comment:
I understand that we're guarding this caller-side (i.e. by making a check that `data/wells` is present before calling this function). However, would generally prefer to make such assumptions/checks directly into the function itself.

As currently defined, this will iterate over every `self.informatics` here and will result in a false negative if we add in a new Informatics instruction that doesn't have ["data"]["wells"]. 

Let's narrow the scope of this function and make it more explicit? E.g. make this function only contain the business logic of comparing two well groups, and shift the `get_wells` call to the caller-side above.

Reply:
Ah. I guess so. good point.